core/src/mender-client.c

```diff
@@ -1069,6 +1081,9 @@ mender_client_update_work_function(void) {
                 if (!mender_update_module->supports_rollback) {
                     mender_log_warning("Rollback not supported for artifacts of type '%s'", mender_update_module->artifact_type);
                     ret = MENDER_FAIL;
+                } else if (aborted_deployment) {
+                    /* Don't rollback if deployment is aborted */
```

Comment:
Wouldn't you as a user expect a rollback if you abort a deployment that isn't committed yet?

Reply:
Right - but the code doesn't support an abort after a reboot - there is nothing publishing the status between reboot and commit, and on commit we've said it's too late. My idea is that before we have rebooted, we should not rollback into the other partition, but just go to failure - if not we'll end up in the wrong partition if we e.g. abort the deployment after the install state has published its status and set the pending image

Reply:
Assume we have `MENDER_UPDATE_STATE_DOWNLOAD -> MENDER_UPDATE_STATE_INSTALL -> MENDER_UPDATE_STATE_REBOOT`.
On each of these states we publish the status and check if the deployment is aborted. If it's aborted when we enter `MENDER_UPDATE_STATE_REBOOT`, then what do we want to do? We haven't rebooted yet as this is caught before the reboot callback is called, so we can either go directly to failure - which deviates from the normal flow, but it does make sense since we haven't actually rebooted, and this is what I originally did, or if we follow the normal state transitions we should go the failure route, which will transition to rollback, but indeed, we have nothing to rollback, which is why I checked for an aborted deployment and set `ret = MENDER_FAIL` in `MENDER_UPDATE_STATE_ROLLBACK`

If we checked if the status is aborted in `MENDER_UPDATE_STATE_COMMIT` (which Lluis said was too late, but I don't quite know why it's too late), then it would make sense to do a proper rollback, as we have rebooted into the new partition, and we would then need to perform a rollback

Reply:
> If it's aborted when we enter `MENDER_UPDATE_STATE_REBOOT`, then what do we want to do? We haven't rebooted yet as this is caught before the reboot callback is called

That brings us to my first reaction to this -- how does it differ from the case where the reboot callback fails? Then we also haven't rebooted yet and that goes to the rollback because that's the failure path, right? I believe the `ROLLBACK` state is responsible for recognizing that it should do nothing. But not by checking `if (aborted_deployment)` but by checking if anything needs to be done to rollback. If not, it should be a no-op.

Reply:
Right - but as of now, it will rollback if I don't check it explicitly, so changing the functionality of rollback seems out of scope for this PR at least, I can remove the check for aborted in rollback, and then we should probably create a separate ticket for a no-op rollback?

Reply:
Works for me! We still have this in the default update module:
```
/* no need for a rollback callback because a reboot without image confirmation is a rollback */
```
so I guess it should just work? Better double-check for sure, but my understanding is that `ROLLBACK` does nothing and then it goes to `ROLLBACK_REBOOT` which reboots the device and goes to `VERIFY_ROLLBACK_REBOOT` and then to `FAILURE`. So if the deployment abortion happens as you described above, it will simply reboot and follow the failure path. No big harm done except that the reboot could be skipped which is what we can fix/improve.